File: python/src/shadow/ltl/formula.py
"""Linear Temporal Logic (LTL) formula AST.

Operators
---------
- ``Atom(pred)``      — atomic proposition, evaluated by a truth assignment
- ``Not(φ)``          — logical negation
- ``And(φ, ψ)``       — logical conjunction
- ``Or(φ, ψ)``        — logical disjunction
- ``Implies(φ, ψ)``   — logical implication (sugar: ¬φ ∨ ψ)
- ``Next(φ)``         — X φ: φ holds at the next state
- ``Until(φ, ψ)``     — φ U ψ: φ holds until ψ becomes true (strong: ψ must occur)
- ``WeakUntil(φ, ψ)`` — φ W ψ: φ holds until ψ, *or* φ holds forever (ψ optional)
- ``Globally(φ)``     — G φ = ¬(⊤ U ¬φ) = φ W ⊥: φ holds at every future state
- ``Finally(φ)``      — F φ = ⊤ U φ: φ holds at some future state

``Atom.pred`` is a string in one of the canonical forms the LTL
checker understands:

    "tool_call:<name>"          — a tool named <name> was called this turn
    "stop_reason:<value>"       — response stop_reason == <value>
    "text_contains:<substring>" — response text contains <substring>
    "true"                      — always true (⊤)
    "false"                     — always false (⊥)

Custom predicates can be injected by passing an ``EvalFn`` to
:func:`~shadow.ltl.checker.check`.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Atom:
    pred: str

    def __str__(self) -> str:
        return self.pred


@dataclass(frozen=True)
class Not:
    child: Formula

    def __str__(self) -> str:
        return f"¬({self.child})"


@dataclass(frozen=True)
class And:
    left: Formula
    right: Formula

    def __str__(self) -> str:
        return f"({self.left} ∧ {self.right})"


@dataclass(frozen=True)
class Or:
    left: Formula
    right: Formula

    def __str__(self) -> str:
        return f"({self.left} ∨ {self.right})"


@dataclass(frozen=True)
class Implies:
    left: Formula
    right: Formula

    def __str__(self) -> str:
        return f"({self.left} → {self.right})"


@dataclass(frozen=True)
class Next:
    child: Formula

    def __str__(self) -> str:
        return f"X({self.child})"


@dataclass(frozen=True)
class Until:
    left: Formula
    right: Formula

    def __str__(self) -> str:
        return f"({self.left} U {self.right})"


@dataclass(frozen=True)
class WeakUntil:
    """Weak until: φ W ψ ≡ (φ U ψ) ∨ G(φ).

    Unlike strong Until, WeakUntil does not require ψ to ever hold —
    if φ holds forever, the formula is satisfied. Used to encode
    "B may only fire after A" without forcing A to actually fire.
    """

    left: Formula
    right: Formula

    def __str__(self) -> str:
        return f"({self.left} W {self.right})"


@dataclass(frozen=True)
class Globally:
    child: Formula

    def __str__(self) -> str:
        return f"G({self.child})"


@dataclass(frozen=True)
class Finally:
    child: Formula

    def __str__(self) -> str:
        return f"F({self.child})"


Formula = Atom | Not | And | Or | Implies | Next | Until | WeakUntil | Globally | Finally


# Convenience constructors for common sub-formulas.


def atom(pred: str) -> Atom:
    return Atom(pred)


def g(phi: Formula) -> Globally:
    return Globally(phi)


def f(phi: Formula) -> Finally:
    return Finally(phi)


def x(phi: Formula) -> Next:
    return Next(phi)


def u(phi: Formula, psi: Formula) -> Until:
    return Until(phi, psi)


def w(phi: Formula, psi: Formula) -> WeakUntil:
    return WeakUntil(phi, psi)


def neg(phi: Formula) -> Not:
    return Not(phi)


def conj(*args: Formula) -> Formula:
    """Conjoin a sequence of formulas (left-associative)."""
    if not args:
        return Atom("true")
    result: Formula = args[0]
    for phi in args[1:]:
        result = And(result, phi)
    return result


def disj(*args: Formula) -> Formula:
    """Disjoin a sequence of formulas (left-associative)."""
    if not args:
        return Atom("false")
    result: Formula = args[0]
    for phi in args[1:]:
        result = Or(result, phi)
    return result


def implies(phi: Formula, psi: Formula) -> Implies:
    return Implies(phi, psi)


# Common shorthands.
TRUE: Atom = Atom("true")
FALSE: Atom = Atom("false")


__all__ = [
    "FALSE",
    "TRUE",
    "And",
    "Atom",
    "Finally",
    "Formula",
    "Globally",
    "Implies",
    "Next",
    "Not",
    "Or",
    "Until",
    "WeakUntil",
    "atom",
    "conj",
    "disj",
    "f",
    "g",
    "implies",
    "neg",
    "u",
    "w",
    "x",
]

File: typescript/src/ltl/formula.ts
/**
 * Linear Temporal Logic (LTLf) formula AST.
 *
 * Discriminated union mirroring the Python `shadow.ltl.formula` types
 * one-to-one. Same operator semantics (finite-trace LTL), same
 * boundary behaviour at the end of the trace, same string forms for
 * atomic predicates ("tool_call:<name>", "stop_reason:<value>",
 * "text_contains:<substr>", "true", "false", "extra:<key>=<value>").
 *
 * The Python AST and this TypeScript AST are kept byte-identical at
 * the JSON level so a formula serialised in either language parses
 * cleanly in the other.
 */

export type Formula =
  | { kind: 'atom'; pred: string }
  | { kind: 'not'; child: Formula }
  | { kind: 'and'; left: Formula; right: Formula }
  | { kind: 'or'; left: Formula; right: Formula }
  | { kind: 'implies'; left: Formula; right: Formula }
  | { kind: 'next'; child: Formula }
  | { kind: 'until'; left: Formula; right: Formula }
  | { kind: 'weakUntil'; left: Formula; right: Formula }
  | { kind: 'globally'; child: Formula }
  | { kind: 'finally'; child: Formula };

export const TRUE: Formula = { kind: 'atom', pred: 'true' };
export const FALSE: Formula = { kind: 'atom', pred: 'false' };

export function atom(pred: string): Formula {
  return { kind: 'atom', pred };
}

export function not(child: Formula): Formula {
  return { kind: 'not', child };
}

export function and(left: Formula, right: Formula): Formula {
  return { kind: 'and', left, right };
}

export function or(left: Formula, right: Formula): Formula {
  return { kind: 'or', left, right };
}

export function implies(left: Formula, right: Formula): Formula {
  return { kind: 'implies', left, right };
}

export function next(child: Formula): Formula {
  return { kind: 'next', child };
}

export function globallyOp(child: Formula): Formula {
  return { kind: 'globally', child };
}

export function finallyOp(child: Formula): Formula {
  return { kind: 'finally', child };
}

export function until(left: Formula, right: Formula): Formula {
  return { kind: 'until', left, right };
}

export function weakUntil(left: Formula, right: Formula): Formula {
  return { kind: 'weakUntil', left, right };
}

export function conj(...args: Formula[]): Formula {
  if (args.length === 0) return TRUE;
  let result = args[0]!;
  for (let i = 1; i < args.length; i++) {
    result = and(result, args[i]!);
  }
  return result;
}

export function disj(...args: Formula[]): Formula {
  if (args.length === 0) return FALSE;
  let result = args[0]!;
  for (let i = 1; i < args.length; i++) {
    result = or(result, args[i]!);
  }
  return result;
}

export function formulaToString(f: Formula): string {
  switch (f.kind) {
    case 'atom':
      return f.pred;
    case 'not':
      return `¬(${formulaToString(f.child)})`;
    case 'and':
      return `(${formulaToString(f.left)} ∧ ${formulaToString(f.right)})`;
    case 'or':
      return `(${formulaToString(f.left)} ∨ ${formulaToString(f.right)})`;
    case 'implies':
      return `(${formulaToString(f.left)} → ${formulaToString(f.right)})`;
    case 'next':
      return `X(${formulaToString(f.child)})`;
    case 'until':
      return `(${formulaToString(f.left)} U ${formulaToString(f.right)})`;
    case 'weakUntil':
      return `(${formulaToString(f.left)} W ${formulaToString(f.right)})`;
    case 'globally':
      return `G(${formulaToString(f.child)})`;
    case 'finally':
      return `F(${formulaToString(f.child)})`;
  }
}
